cartography/intel/okta/oktaintel.py

```diff
@@ -0,0 +1,1193 @@
+
+    app_props["features"] = okta_application.features
+    app_props["sign_on_mode"] = okta_application.signOnMode
+
+    return app_props
+
+
+def _load_okta_applications(neo4j_session, okta_org_id, app_list, okta_update_tag):
+    """
+    Add application into the graph
+    :param neo4j_session: session with the Neo4j server
+    :param okta_org_id: okta organization id
+    :param app_list: application list - Array of dictionary
+    :param okta_update_tag: The timestamp value to set our new Neo4j resources with
+    :return: Nothing
+    """
+    ingest_statement = """
+    MATCH (org:OktaOrganization{id: {ORG_ID}})
+    WITH org
+    UNWIND {APP_LIST} as app_data
+    MERGE (new_app:OktaApplication{id: app_data.id})
+    ON CREATE SET new_app.firstseen = timestamp()
+    SET new_app.name = app_data.name,
+    new_app.label = app_data.label,
+    new_app.created = app_data.created,
+    new_app.okta_last_updated = app_data.okta_last_updated,
+    new_app.status = app_data.status,
+    new_app.activated = app_data.activated,
+    new_app.features = app_data.features,
+    new_app.sign_on_mode = app_data.sign_on_mode,
+    new_app.lastupdated = {okta_update_tag}
+    WITH org, new_app
+    MERGE (org)-[org_r:RESOURCE]->(new_app)
+    ON CREATE SET org_r.firstseen = timestamp()
+    SET org_r.lastupdated = {okta_update_tag}
+    """
+
+    neo4j_session.run(
+        ingest_statement,
+        ORG_ID=okta_org_id,
+        APP_LIST=app_list,
+        okta_update_tag=okta_update_tag,
+    )
+
+
+def _get_application_assigned_users(api_client, app_id):
+    """
+    Get users assigned to a specific application
+    :param api_client: api client
+    :param app_id: application id to get users from
+    :return: Array of user id
+    """
+    app_users = []
+
+    next_url = None
+    while True:
+        try:
+            # https://developer.okta.com/docs/reference/api/apps/#list-users-assigned-to-application
+            if next_url:
+                paged_response = api_client.get(next_url)
+            else:
+                params = {
+                    'limit': 500,
+                }
+                paged_response = api_client.get_path(f'/{app_id}/users', params)
+        except OktaError as okta_error:
+            logger.debug(f"Got error while going through list application assigned users {okta_error}")
+            break
+
+        for user_id in transform_application_users(paged_response.text):
+            app_users.append(user_id)
+
+        if not _is_last_page(paged_response):
+            next_url = paged_response.links.get("next").get("url")
+        else:
+            break
+
+    return app_users
+
+
+def transform_application_users(json_app_data):
+    """
+    Transform application users data for graph consumption
+    :param json_app_data: raw json application data
+    :return: individual user id as yield
+    """
+
+    app_data = json.loads(json_app_data)
+    for user in app_data:
+        yield user["id"]
```

Comment:
Why is yield being used here? Is it necessary for this to be a generator function?

Reply:
This applies to other uses of yield in this module as well.